clippy_lints/src/methods/map_flatten.rs

```diff
@@ -68,16 +70,46 @@ pub(super) fn check<'tcx>(
         },
     };
 
-    let func_snippet = snippet(cx, map_arg.span, "..");
-    let hint = format!(".and_then({})", func_snippet);
+    let mut applicability = Applicability::MachineApplicable;
+    let sugg_span = expr.span.with_lo(recv.span.hi());
     let lint_info = format!("called `map(..).flatten()` on an `{}`", caller_type);
-    span_lint_and_sugg(
-        cx,
-        MAP_FLATTEN,
-        expr.span.with_lo(recv.span.hi()),
-        &lint_info,
-        "try using `and_then` instead",
-        hint,
-        Applicability::MachineApplicable,
-    );
+    let help_msgs = [
+        "try replacing `map` with `and_then` instead".to_string(),
+        "and remove the `.flatten()`".to_string(),
+    ];
+    let suggestion = &Sugg::hir_with_applicability(cx, map_arg, "..", &mut applicability);
+    let hint = format!(".and_then({})", suggestion);
+    span_lint_and_emitted_sugg(cx, MAP_FLATTEN, sugg_span, &lint_info, &help_msgs, hint, applicability);
+}
+
+fn span_lint_and_emitted_sugg<'a, T>(
+    cx: &'a T,
+    lint: &'static Lint,
+    sp: Span,
+    msg: &str,
+    help_msgs: &[String],
+    sugg: String,
+    applicability: Applicability,
+) where
+    T: LintContext,
+{
+    const MAX_LINES: usize = rustc_errors::emitter::MAX_SUGGESTION_HIGHLIGHT_LINES;
+    let lines_count = sugg.lines().count();
+    if lines_count < MAX_LINES {
+        if let Some(help) = help_msgs.first() {
+            span_lint_and_sugg(cx, lint, sp, msg, help, sugg, applicability);
+        }
+    } else {
+        if help_msgs.len() == 2 {
+            let lines = sugg.split("\n").collect::<Vec<_>>();
+            let top_part = lines[..MAX_LINES / 2].join("\n");
+            let bottom_part = lines[lines_count - MAX_LINES / 2..].join("\n");
+            //println!("top: {}\nbottom: {}", top_part, bottom_part);
+            span_lint_and_then(cx, lint, sp, msg, |diag| {
+                // Error: "Cannot replace slice of data that was already replaced" (caused by overlapping Span I think?)
```

Comment:
rustfix doesn't work for multipart suggestions like this one. In these suggestions we usually split up the tests into a `_fixable` and a normal test